src/worker/environment.rs

```diff
@@ -17,6 +20,9 @@ pub struct Environment {
     pub storage: Arc<Storage>,
     pub connection_pool: DieselPool,
     pub emails: Emails,
+
+    /// A lazily initialised cache of the most popular crates ready to use in typosquatting checks.
+    typosquat_cache: OnceLock<Result<typosquat::Cache, typosquat::CacheError>>,
```

Comment:
if I understand this implementation correctly, if the first attempt to fill the cache fails it will not be tried again?

I noticed a similar situation with the `repository` field of this struct lately and fixed it like this: https://github.com/rust-lang/crates.io/blob/f8fc9ae18efa59f4965ce2ce07de4250ccdb74e0/src/worker/environment.rs#L40-L54

basically, for each access of the method it checks whether the field is initialized and returns the initialized thing if so. otherwise it tries to initialize and returns an error if it fails. the next time the method is used the initialization will be retried though.

I assume this might also resolve the need for some of the `Arc` gymnastics in the error type below :)

Reply:
> if I understand this implementation correctly, if the first attempt to fill the cache fails it will not be tried again?

Yep. That was actually intentional on my part — the typosquatting checks aren't mission critical, so I'd rather it fails with a warning and just disables itself until the next dyno restart rather than putting load on the database trying to rebuild the cache each time a crate is published if there's some sort of weird problem preventing the query from running properly.

I'm open minded on whether the right thing to do is to change it to be more like the repo behaviour your mentioned, or if I should just document that choice more clearly in the code. Thoughts?

Reply:
okay, if it's intentional then that's fine. just wanted to check :)